File: simulation-car/src/main/scala/model/inspector/RoadSimStatistics.scala
package model.inspector

import model.car.{ Agent, CarAgent }
import model.simulation.{ ModelSimulationListener, SimulationInspector }

trait RoadSimStatistics extends ModelSimulationListener:
  def currentAverageSpeed: Double

  def averageSpeed: Double

  def minSpeed: Double

  def maxSpeed: Double

object RoadSimStatistics:

  def apply(): RoadSimStatistics = RoadSimStatisticsImpl()

  private case class RoadSimStatisticsImpl() extends RoadSimStatistics:
    private var _currentAverageSpeed: Double = .0
    private var _averageSpeed: Double = .0
    private var _minSpeed: Double = .0
    private var _maxSpeed: Double = .0

    private def currentAverageSpeed_=(n: Double): Unit = _currentAverageSpeed = n
    private def averageSpeed_=(n: Double): Unit = _averageSpeed = n
    private def minSpeed_=(n: Double): Unit = _minSpeed = n
    private def maxSpeed_=(n: Double): Unit = _maxSpeed = n
    private def log(msg: String): Unit = println("[STAT] " + msg)

    override def currentAverageSpeed: Double = _currentAverageSpeed
    override def averageSpeed: Double = _averageSpeed
    override def minSpeed: Double = _minSpeed
    override def maxSpeed: Double = _maxSpeed


    override def notifyInit(t: Int, simulation: SimulationInspector): Unit =
      currentAverageSpeed = 0d
      averageSpeed = 0d

    override def notifyStepDone(t: Int, simulation: SimulationInspector): Unit =
      currentAverageSpeed = 0
      maxSpeed = -1d
      minSpeed = Double.MaxValue
      val agents: List[Agent] = simulation.agents()

      agents.foreach(agent => {
        val car: CarAgent = agent.asInstanceOf[CarAgent]
        val currSpeed = car.currentSpeed
        currentAverageSpeed = currentAverageSpeed + currSpeed
        if (currSpeed > maxSpeed) maxSpeed = currSpeed
        else if (currSpeed < minSpeed) minSpeed = currSpeed
      })

      if (agents.nonEmpty) currentAverageSpeed = currentAverageSpeed / agents.size
      log("average speed: " + currentAverageSpeed)

    override def notifyEnd(simulation: SimulationInspector): Unit = {}

File: simulation-car/src/main/java/inspector/RoadSimStatistics.java
package inspector;

import actors.CarActor;
import akka.actor.typed.ActorRef;
import simulation.InspectorSimulation;
import simulation.listener.ModelSimulationListener;

import java.util.List;

/**
 * Simple class keeping track of some statistics about a traffic simulation
 * - average number of cars
 * - min speed
 * - max speed
 */
public class RoadSimStatistics implements ModelSimulationListener {
    private double currentAverageSpeed;
    private double averageSpeed;
    private double minSpeed;
    private double maxSpeed;

    public RoadSimStatistics() {
    }

    public void updateAverageSpeed(final double currSpeed, final int totalActors) {
        this.currentAverageSpeed += currSpeed;
        if (currSpeed > this.maxSpeed) {
            this.maxSpeed = currSpeed;
        } else if (currSpeed < this.minSpeed) {
            this.minSpeed = currSpeed;
        }

        if (totalActors > 0) {
            this.currentAverageSpeed /= totalActors;
        }
    }

    public double currentAverageSpeed() {
        return this.currentAverageSpeed;
    }

    public double averageSpeed() {
        return this.averageSpeed;
    }

    public double minSpeed() {
        return this.minSpeed;
    }

    public double maxSpeed() {
        return this.maxSpeed;
    }

    @Override
    public void notifyInit(final int t, final InspectorSimulation simulation) {
        // log("reset: " + t);
        this.averageSpeed = 0;
        this.currentAverageSpeed = 0;
    }

    @Override
    public void notifyStepDone(final int t, final InspectorSimulation simulation) {
        this.currentAverageSpeed = 0;

        this.maxSpeed = -1;
        this.minSpeed = Double.MAX_VALUE;
        final List<ActorRef<CarActor.Command>> actors = simulation.actors();
//        actors.forEach(actor -> actor.tell(CarActor.UpdateStats.apply(this)));

//        for (final var agent : actors) {
//            final CarAgent car = (CarAgent) agent;
//            final double currSpeed = car.getCurrentSpeed();
//            this.currentAverageSpeed += currSpeed;
//            if (currSpeed > this.maxSpeed) {
//                this.maxSpeed = currSpeed;
//            } else if (currSpeed < this.minSpeed) {
//                this.minSpeed = currSpeed;
//            }
//        }
//
//        if (!actors.isEmpty()) {
//            this.currentAverageSpeed /= actors.size();
//        }
//        this.log("average speed: " + this.currentAverageSpeed);
    }

    @Override
    public void notifyEnd(final InspectorSimulation simulation) {

    }

    private void log(final String msg) {
        System.out.println("[STAT] " + msg);
    }

}
